### ISSUE TITLE
When I use Densenet as backbone, model can't detect  any targets.

### ISSUE THREAD

**Issue by MPolaris** [open-mmlab/mmdetection#2495]
I have a research program for Mammogram, I plan use Cascade RCNN as base frame.
First, I use the official config(cascade_rcnn_r50_fpn_1x.py) and official pretrained model, model work very well. 
Then, I implement DenseNet and use it as a backbone, only change the backbone config. During training, every thing is normal, loss_rpn_cls , loss_bbox, loss etc...  
Unique abnormal is acc ,it will be reach 99 quickly, three stage are. On contrast, resenet is between 93 to 96.
So any suggestion for me? How can I find out the problem？
Thanks for your help!
### Densenet.py
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from collections import OrderedDict

from mmcv.runner import load_checkpoint
from mmdet.utils import get_root_logger
from ..registry import BACKBONES

class _DenseLayer(nn.Sequential):

    def __init__(self, num_input_features, growth_rate, bn_size, drop_rate):
        super(_DenseLayer, self).__init__()
        self.add_module('norm1', nn.BatchNorm2d(num_input_features)),
        self.add_module('relu1', nn.ReLU(inplace=True)),
        self.add_module('conv1', nn.Conv2d(num_input_features, bn_size *
                        growth_rate, kernel_size=1, stride=1, bias=False)),
        self.add_module('norm2', nn.BatchNorm2d(bn_size * growth_rate)),
        self.add_module('relu2', nn.ReLU(inplace=True)),
        self.add_module('conv2', nn.Conv2d(bn_size * growth_rate, growth_rate,
                        kernel_size=3, stride=1, padding=1, bias=False)),
        self.drop_rate = drop_rate

    def forward(self, x):
        new_features = super(_DenseLayer, self).forward(x)
        if self.drop_rate > 0:
            new_features = F.dropout(new_features, p=self.drop_rate, training=self.training)
        return torch.cat([x, new_features], 1)
class _DenseBlock(nn.Sequential):

    def __init__(self, num_layers, num_input_features, bn_size, growth_rate, drop_rate):
        super(_DenseBlock, self).__init__()
        for i in range(num_layers):
            layer = _DenseLayer(num_input_features + i * growth_rate, growth_rate, bn_size, drop_rate)
            self.add_module('denselayer%d' % (i + 1), layer)

class _Transition(nn.Sequential):

    def __init__(self, num_input_features, num_output_features):
        super(_Transition, self).__init__()
        self.add_module('norm', nn.BatchNorm2d(num_input_features))
        self.add_module('relu', nn.ReLU(inplace=True))
        self.add_module('conv', nn.Conv2d(num_input_features, num_output_features,
                                          kernel_size=1, stride=1, bias=False))
        self.add_module('pool', nn.AvgPool2d(kernel_size=2, stride=2))

@BACKBONES.register_module
class DenseNet(nn.Module):
    # num_init_features, growth_rate, block_config
    arch_settings = {
        121: (64, 32, (6, 12, 24, 16)),
        161: (96, 48, (6, 12, 36, 24)),
        169: (64, 32, (6, 12, 32, 32)),
        201: (64, 32, (6, 12, 48, 32)),
    }

    def __init__(self, depth=121,
                 num_stages=4,
                 frozen_stages=-1,
                 norm_eval=True,
                 bn_size=4,
                 drop_rate=0
                 ):
        super(DenseNet, self).__init__()
        self.num_stages = num_stages
        assert num_stages >= 1 and num_stages <= 4
        self.frozen_stages = frozen_stages
        self.norm_eval = norm_eval

        num_init_features, growth_rate, block_config = self.arch_settings[depth]

        # First convolution
        self.features = nn.Sequential(OrderedDict([
            ('conv0', nn.Conv2d(3, num_init_features, kernel_size=7, stride=2, padding=3, bias=False)),
            ('norm0', nn.BatchNorm2d(num_init_features)),
            ('relu0', nn.ReLU(inplace=True)),
            ('pool0', nn.MaxPool2d(kernel_size=3, stride=2, padding=1)),
        ]))

        # Each denseblock
        num_features = num_init_features
        for i, num_layers in enumerate(block_config):
            block = _DenseBlock(num_layers=num_layers, num_input_features=num_features,
                                bn_size=bn_size, growth_rate=growth_rate, drop_rate=drop_rate)
            self.features.add_module('denseblock%d' % (i + 1), block)
            num_features = num_features + num_layers * growth_rate
            if i != len(block_config) - 1:
                trans = _Transition(num_input_features=num_features, num_output_features=num_features // 2)
                self.features.add_module('transition%d' % (i + 1), trans)
                num_features = num_features // 2

        # Final batch norm
        self.features.add_module('norm5', nn.BatchNorm2d(num_features))

        # # Linear layer
        # self.classifier = nn.Linear(num_features, num_classes)

        # Official init from torch repo.
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.Linear):
                nn.init.constant_(m.bias, 0)

        self._freeze_stages()

    def _freeze_stages(self):
        # 'Not implemented yet.'
        return

    def init_weights(self, pretrained=None):
        if isinstance(pretrained, str):
            logger = get_root_logger()
            load_checkpoint(self, pretrained, strict=False, logger=logger, isolddensenetload=True)
        elif pretrained is None:
            return
        else:
            raise TypeError('pretrained must be a str or None')

   def forward(self, x,train = False):
        outs = []
        features = self.features.conv0(x)
        features = self.features.norm0(features)
        features = self.features.relu0(features)
        features = self.features.pool0(features)
        features = self.features.denseblock1(features)
        features = self.features.transition1(features)
        outs.append(features)
        features = self.features.denseblock2(features)
        features = self.features.transition2(features)
        outs.append(features)
        features = self.features.denseblock3(features)
        features = self.features.transition3(features)
        outs.append(features)
        features = self.features.denseblock4(features)
        features = self.features.norm5(features)
        outs.append(features)
        return tuple(outs)


    def train(self, mode=True):
        super(DenseNet, self).train(mode)
        self._freeze_stages()
        if mode and self.norm_eval:
            for m in self.modules():
                # trick: eval have effect on BatchNorm only
                if isinstance(m, nn.BatchNorm2d):
                    for param in m.parameters():
                        param.requires_grad = True
                    m.eval()
```
### config.py
```python
# model settings
model = dict(
    type='CascadeRCNN',
    num_stages=3,
    pretrained='/models/densenet/densenet121-a639ec97.pth',
    backbone=dict(
        type='DenseNet',
        depth=121,
        num_stages=4),
    neck=dict(
        type='FPN',
        in_channels=[128, 256, 512, 1024],
        out_channels=256,
        num_outs=5),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_scales=[8],
        anchor_ratios=[0.5, 1.0, 2.0],
        anchor_strides=[4, 8, 16, 32, 64],
        target_means=[.0, .0, .0, .0],
        target_stds=[1.0, 1.0, 1.0, 1.0],
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
        loss_bbox=dict(type='SmoothL1Loss', beta=1.0 / 9.0, loss_weight=1.0)),
    bbox_roi_extractor=dict(
        type='SingleRoIExtractor',
        roi_layer=dict(type='RoIAlign', out_size=7, sample_num=2),
        out_channels=256,
        featmap_strides=[4, 8, 16, 32]),
    bbox_head=[
        dict(
            type='SharedFCBBoxHead',
            num_fcs=2,
            in_channels=256,
            fc_out_channels=1024,
            roi_feat_size=7,
            num_classes=3,
            target_means=[0., 0., 0., 0.],
            target_stds=[0.1, 0.1, 0.2, 0.2],
            reg_class_agnostic=True,
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0)),
        dict(
            type='SharedFCBBoxHead',
            num_fcs=2,
            in_channels=256,
            fc_out_channels=1024,
            roi_feat_size=7,
            num_classes=3,
            target_means=[0., 0., 0., 0.],
            target_stds=[0.05, 0.05, 0.1, 0.1],
            reg_class_agnostic=True,
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0)),
        dict(
            type='SharedFCBBoxHead',
            num_fcs=2,
            in_channels=256,
            fc_out_channels=1024,
            roi_feat_size=7,
            num_classes=3,
            target_means=[0., 0., 0., 0.],
            target_stds=[0.033, 0.033, 0.067, 0.067],
            reg_class_agnostic=True,
            loss_cls=dict(
                type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
            loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0))
    ])
# model training and testing settings
train_cfg = dict(
    rpn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.7,
            neg_iou_thr=0.3,
            min_pos_iou=0.3,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.5,
            neg_pos_ub=-1,
            add_gt_as_proposals=False),
        allowed_border=0,
        pos_weight=-1,
        debug=False),
    rpn_proposal=dict(
        nms_across_levels=False,
        nms_pre=2000,
        nms_post=2000,
        max_num=2000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=[
        dict(
            assigner=dict(
                type='MaxIoUAssigner',
                pos_iou_thr=0.5,
                neg_iou_thr=0.5,
                min_pos_iou=0.5,
                ignore_iof_thr=-1),
            sampler=dict(
                type='RandomSampler',
                num=512,
                pos_fraction=0.25,
                neg_pos_ub=-1,
                add_gt_as_proposals=True),
            pos_weight=-1,
            debug=False),
        dict(
            assigner=dict(
                type='MaxIoUAssigner',
                pos_iou_thr=0.6,
                neg_iou_thr=0.6,
                min_pos_iou=0.6,
                ignore_iof_thr=-1),
            sampler=dict(
                type='RandomSampler',
                num=512,
                pos_fraction=0.25,
                neg_pos_ub=-1,
                add_gt_as_proposals=True),
            pos_weight=-1,
            debug=False),
        dict(
            assigner=dict(
                type='MaxIoUAssigner',
                pos_iou_thr=0.7,
                neg_iou_thr=0.7,
                min_pos_iou=0.7,
                ignore_iof_thr=-1),
            sampler=dict(
                type='RandomSampler',
                num=512,
                pos_fraction=0.25,
                neg_pos_ub=-1,
                add_gt_as_proposals=True),
            pos_weight=-1,
            debug=False)
    ],
    stage_loss_weights=[1, 0.5, 0.25])
test_cfg = dict(
    rpn=dict(
        nms_across_levels=False,
        nms_pre=1000,
        nms_post=1000,
        max_num=1000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        score_thr=0.05, nms=dict(type='soft_nms', iou_thr=0.15), max_per_img=100))
# dataset settings
dataset_type = 'underwaterdataloader'
data_root = '../data/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1228, 1024), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1228, 1024),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    imgs_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/trainV1.json',
        img_prefix=data_root + 'images/',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/valV1.json',
        img_prefix=data_root + 'images/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/valV1.json',
        img_prefix=data_root + 'images/',
        pipeline=test_pipeline))
evaluation = dict(interval=1, metric='bbox')
# optimizer
optimizer = dict(type='SGD', lr=data['imgs_per_gpu']*0.00125*1, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=200,
    warmup_ratio=1.0 / 3,
    step=[8, 11])
checkpoint_config = dict(interval=1)
# yapf:disable
log_config = dict(
    interval=20,
    hooks=[
        dict(type='TextLoggerHook'),
        # dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
# runtime settings
total_epochs = 12
dist_params = dict(backend='nccl')
log_level = 'INFO'
work_dir = data_root + 'work_dirs/densenet_imgnet/'
load_from = None
resume_from = None
workflow = [('train', 1)]
```

**Comment by FancyRay** [comment]
Hi, I also tried to use densent as the backbone of maskrcnn.
But the error occured as follows. Is there anything wrong with fpn? Do you have any suggestion?
Thanks a lot.
![image](https://user-images.githubusercontent.com/55759544/79944622-99e76780-849e-11ea-92c3-4cfebeed7235.png)

**Comment by FancyRay** [comment]
Hi, I also tried to use densent as the backbone of maskrcnn.
But the error occured as follows. Is there anything wrong with fpn? Do you have any suggestion?
Thanks a lot.
<img width="454" alt="Snipaste_2020-04-22_13-40-27" src="https://user-images.githubusercontent.com/55759544/79944727-dc10a900-849e-11ea-95dc-8c9fd17f62da.png">

**Comment by FancyRay** [comment]
Hi, I also tried to use densenet as the backbone of maskrcnn. But the loaded state dict does not match. I wonder how you solve the problem.
Thanks a lot.

**Comment by MPolaris** [comment]
> Hi, I also tried to use densent as the backbone of maskrcnn.
> But the error occured as follows. Is there anything wrong with fpn? Do you have any suggestion?
> Thanks a lot.
> ![image](https://user-images.githubusercontent.com/55759544/79944622-99e76780-849e-11ea-92c3-4cfebeed7235.png)

First, you should know the channels of outputs for your backbone, and change the neck's 'in_channels' property correspondingly.
It's easy get the channels, for example:
```python
if __name__ == '__main__':
    model = DenseNet()
    img = torch.randn(2, 3, 299, 299)
    res = model(img)
    for r in res:
        print(r.shape[1])
```

**Comment by MPolaris** [comment]
> Hi, I also tried to use densenet as the backbone of maskrcnn. But the loaded state dict does not match. I wonder how you solve the problem.
> Thanks a lot.

You should rewrite the function 'init_weights'.
If your Densenet from Pytorch's lib, and for some version reason, the state_dict name can't match and should be rename before load.
Here is the code you can reference:
```python
    import re
    checkpoint = torch.load('Pytorch lib model')
    pattern = re.compile(r'^(.*denselayer\d+\.(?:norm|relu|conv))\.((?:[12])\.(?:weight|bias|running_mean|running_var))$')
    for key in list(checkpoint.keys()):
        res = pattern.match(key)
        if res:
            new_key = res.group(1) + res.group(2)
            checkpoint[new_key] = checkpoint[key]
            del checkpoint[key]
    model.load_statedict(checkpoint)
```

**Comment by zzzjoey** [comment]
After changing the backbone to densenet, I encountered the same problem. Have you solved it?

**Comment by zzzjoey** [comment]
> After changing the backbone to densenet, I encountered the same problem. Have you solved it?

I think I found a solution. After making sure there was no problem with the data, I found that there were many problems with the weight names when loading the pre-trained weights, like missing keys and unexpected key. After modifying them, I was able to load the pre-trained weights normally and detect the box during testing.
